python/cudf/cudf/core/udf/udf_kernel_base.py

```diff
@@ -0,0 +1,193 @@
+        """
+        Get a dict of globals needed to exec the kernel
+        string.
+        """
+
+    def _construct_signature(self, return_type):
+        """
+        Build the signature of numba types that will be used to
+        actually JIT the kernel itself later, accounting for types
+        and offsets. Skips columns with unsupported dtypes.
+        """
+        if not return_type.is_internal:
+            return_type = CPointer(return_type)
+        else:
+            return_type = return_type[::1]
+        # Tuple of arrays, first the output data array, then the mask
+        return_type = Tuple((return_type, boolean[::1]))
+        offsets = []
+        sig = [return_type, int64]
+        for col in _supported_cols_from_frame(self.frame).values():
+            sig.append(_masked_array_type_from_col(col))
+            offsets.append(int64)
+
+        # return_type, size, data, masks, offsets, extra args
+        sig = void(*(sig + offsets + [typeof(arg) for arg in self.args]))
+
+        return sig
+
+    @_performance_tracking
+    def _get_udf_return_type(self):
+        # present a row containing all fields to the UDF and try and compile
+        compile_sig = (self.frame_type, *(typeof(arg) for arg in self.args))
+
+        # Get the return type. The PTX is also returned by compile_udf, but is not
+        # needed here.
+        with _CUDFNumbaConfig():
+            ptx, output_type = cudautils.compile_udf(self.func, compile_sig)
+
+        if not isinstance(output_type, MaskedType):
+            numba_output_type = numpy_support.from_dtype(np.dtype(output_type))
+        else:
+            numba_output_type = output_type
+
+        result = (
+            numba_output_type
+            if not isinstance(numba_output_type, MaskedType)
+            else numba_output_type.value_type
+        )
+        result = result if result.is_internal else result.return_type
+
+        # _get_udf_return_type will throw a TypingError if the user tries to use
+        # a field in the row containing an unsupported dtype, except in the
+        # edge case where all the function does is return that element:
+
+        # def f(row):
+        #    return row[<bad dtype key>]
+        # In this case numba is happy to return MaskedType(<bad dtype key>)
+        # because it relies on not finding overloaded operators for types to raise
+        # the exception, so we have to explicitly check for that case.
+        if isinstance(result, Poison):
+            raise TypingError(str(result))
+
+        return result
+
+    def compile_kernel(self):
+        """
+        Compile the kernel and return it as well as the
+        return type.
+        """
+
+        # First compilation pass compiles the UDF alone
+        # gets us the return type to allocate the output
+        # and determines if NRT must be enabled
+        nrtctx = NRTContext()
+        with nrtctx:
+            return_type = self._get_udf_return_type()
```

Comment:
This feels like an abuse of a context manager. I had to read the code a few times to understand what is happening, and if I'm understanding correctly, we enter the `NRTContext` solely so that somewhere down the call stack we can hit the code in `masked_typing` that recognizes that we are inside this context and therefore sets the context's `use_nrt` flag so that we can pass it into `compile_kernel_string` below. That feels very nonlocal and hard to track. Can we instead either return `use_nrt` from `_get_udf_return_type`, or if not return that directly, return some other metadata that can be used in this function to indicate that we need NRT while ignoring it in other places where `_get_udf_return_type` is called but we don't need NRT?

Reply:
We need some kind of proxy for if NRT is needed, here the idea is to use the fact that type we _know_ certain types require NRT to be instantiated as that proxy (the condition inside the masked constructor is that the type being considered `has_nrt_meminfo`). Without that its hard to come up with another way of statically knowing if managed types are being used in the code, any such analysis would require propagating typing through the full code anyways. Theoretically we there's lowering we're doing that's extra work to make the determination, but right now numba's typing and lowering passes don't really decouple in a way that we can just run the typing part here.

Reply:
All that said, I'm open to ideas of course!

Reply:
We discussed this offline. Now that I understand the problem better, I agree that we probably can't avoid some degree of action at a distance here since most of the detection is tied up deep in numba's compilation stack. We're going to try and simplify the code a bit by merging context managers and adding some comments, but that's as much as we can likely do for now.